stdlib/json.gr

```diff
@@ -0,0 +1,2103 @@
+    Buffer.addString(s, buffer)
+    None
+  } else {
+    from WasmF64 use { (<) }
+    // JSON standard doesn't allow NaN or infinite values in numbers,
+    // but WASM f64 (IEEE 754-2008), as well as Grain's number types do
+    // (Float64 as well as Number). This is the only reason that the
+    // formatting needs to return a Result and not just a String
+    // directly. Other possible choices were to throw exceptions or to
+    // continue formatting without representing these values correctly
+    // (like JavaScript's JSON.stringify).
+    if (isNaN(value)) {
+      Some(InvalidNumber("NaN is not allowed in JsonNumber"))
+    } else if (value < 0.0W) {
+      Some(InvalidNumber("-Infinity is not allowed in JsonNumber"))
+    } else {
+      Some(InvalidNumber("Infinity is not allowed in JsonNumber"))
+    }
+  }
+}
+
+@unsafe
+let printNumber = (value: Number, buffer: Buffer.Buffer) => {
+  from WasmI32 use { (&), (==), (!=), (<<), (>>) }
+
+  let ptr = WasmI32.fromGrain(value)
+  let ret = if ((ptr & 1n) != 0n) {
+    printNumberWasmI32(untagSimpleNumber(value), buffer)
+    None
+  } else if ((ptr & 7n) == Tags._GRAIN_GENERIC_HEAP_TAG_TYPE) {
+    let tag = WasmI32.load(ptr, 0n)
+    match (tag) {
+      t when t == Tags._GRAIN_BOXED_NUM_HEAP_TAG => {
+        let numberTag = WasmI32.load(ptr, 4n)
+        match (numberTag) {
+          t when t == Tags._GRAIN_INT64_BOXED_NUM_TAG => {
+            let asWasmI64 = WasmI64.load(ptr, _INT64_BOXED_VALUE_OFFSET)
+            printNumberWasmI64(asWasmI64, buffer)
+            None
+          },
+          t when t == Tags._GRAIN_BIGINT_BOXED_NUM_TAG => {
+            Buffer.addString(BI.bigIntToString10(ptr), buffer)
+            None
+          },
+          t when t == Tags._GRAIN_RATIONAL_BOXED_NUM_TAG => {
+            // JSON does not support rationals as a compromise
+            // we coerce them to an f64 and print that
+            // this means there is a slight loss in precision
+            let asFloat64 = Numbers.coerceNumberToFloat64(value)
+            let ptr = WasmI32.fromGrain(asFloat64)
+            let asWasmF64 = WasmF64.load(ptr, _Float64_BOXED_VALUE_OFFSET)
+            printNumberWasmF64(asWasmF64, buffer)
+          },
+          t when t == Tags._GRAIN_FLOAT64_BOXED_NUM_TAG => {
+            let asWasmF64 = WasmF64.load(ptr, _Float64_BOXED_VALUE_OFFSET)
+            printNumberWasmF64(asWasmF64, buffer)
+          },
+          _ => {
+            throw UnknownNumberTag
+          },
+        }
+      },
+      _ => {
+        throw UnknownNumberTag
+      },
+    }
+  } else {
+    throw UnknownNumberTag
+  }
+  ignore(value)
+  ret
+}
+
+// Note that this could relatively easily be integrated with
+// printElementCompact to avoid code duplication, but it would mean
+// compromising peak performance and also increasing complexity in other ways.
+let rec printElementPretty =
+  (
+    json: Json,
+    implHelper: JsonWriterConfig,
```

Comment:
Is this the best name now? Should it just be `writer`?

Reply:
I switched it to `printElement` as I think thats more descriptive then `writer`.